tests/wpt_tests.cpp

```diff
@@ -0,0 +1,250 @@
+        std::cout << "     output: null" << std::endl;
+      }
+    }
+  }
+  TEST_SUCCEED()
+}
+
+bool urltestdata_encoding() {
+
+  TEST_START()
+  ondemand::parser parser;
+
+  RUN_TEST(file_exists(URLTESTDATA_JSON));
+  padded_string json = padded_string::load(URLTESTDATA_JSON);
+  std::cout << "  loaded " << URLTESTDATA_JSON << " (" << json.size() << " kB)"
+            << std::endl;
+  ondemand::document doc = parser.iterate(json);
+  for (auto element : doc.get_array()) {
+    if (element.type() == ondemand::json_type::string) {
+      std::cout << "     comment: " << element.get_string() << std::endl;
+    } else if (element.type() == ondemand::json_type::object) {
+      ondemand::object object = element.get_object();
+      auto input_element = object["input"];
+      std::string_view input;
+      auto error = input_element.get_string().get(input);
+      if (!error) {
+        std::cout << "     input " << input << std::endl;
+      } else {
+        // we have a non-UTF-8 input.
+        input = input_element.raw_json_token();
+        std::cout << "     raw input " << input << std::endl;
+        // raw input is a quoted string, unescaped.
+        std::cout << "    warning: invalid UTF-8 input!" << std::endl;
+      }
+      std::string_view base;
+      if (!object["base"].get(base)) {
+        std::cout << "     base " << base << std::endl;
+      }
+      bool failure = false;
+      if (!object["failure"].get(failure)) {
+        std::cout << "     failure" << std::endl;
+      } else {
+        std::string_view href = object["href"];
+        std::cout << "     href " << href << std::endl;
+
+        std::string_view origin;
+        if (!object["     origin"].get(origin)) {
+          // origin is optional.
+          std::cout << "     origin " << origin << std::endl;
+        }
+
+        std::string_view protocol = object["protocol"];
+        std::cout << "     protocol " << protocol << std::endl;
+
+        std::string_view username = object["username"];
+        std::cout << "     username " << username << std::endl;
+
+        std::string_view password = object["password"];
+        std::cout << "     password " << password << std::endl;
+
+        std::string_view host = object["host"];
+        std::cout << "     host " << host << std::endl;
+
+        std::string_view hostname = object["hostname"];
+        std::cout << "     hostname " << hostname << std::endl;
+
+        std::string_view port = object["port"];
+        std::cout << "     port " << port << std::endl;
+
+        std::string_view search = object["search"];
+        std::cout << "     search " << search << std::endl;
+        std::string_view hash = object["hash"];
+        std::cout << "     hash " << hash << std::endl;
+      }
+    }
+  }
+  TEST_SUCCEED()
+}
+
+int main() {
+  std::cout << "Running WPT tests.\n" << std::endl;
+
+  if (percent_encoding() && setters_tests_encoding() && toascii_encoding() &&
```

Comment:
I might be wrong here but: I believe this does not run `setters_tests_encoding()` if `percent_encoding()` fails. Can we make sure each test runs regardless?

Reply:
If you'd like. Changed by the next commit.